File: backend/app/schemas/instrument.py
"""Instrument schemas."""

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.models.enums import InstrumentType, Timeframe


class QuoteResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    instrument_id: uuid.UUID
    last_price: Decimal
    previous_close: Decimal | None = None
    change: Decimal | None = None
    change_pct: Decimal | None = None
    day_open: Decimal | None = None
    day_high: Decimal | None = None
    day_low: Decimal | None = None
    volume: int | None = None
    updated_at: datetime


class SignalStats(BaseModel):
    total_signals: int = 0
    bullish: int = 0
    bearish: int = 0
    confirmed: int = 0
    invalidated: int = 0


class InstrumentListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    symbol: str
    name: str
    instrument_type: InstrumentType
    exchange: str
    currency: str
    sector_name: str | None = None


class PaginatedInstruments(BaseModel):
    items: list[InstrumentListItem]
    total: int
    limit: int
    offset: int


class CandleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    timeframe: Timeframe
    ts: datetime
    open: Decimal
    high: Decimal
    low: Decimal
    close: Decimal
    volume: int | None = None


class PaginatedCandles(BaseModel):
    items: list[CandleResponse]
    timeframe: Timeframe
    limit: int
    has_more: bool = False


class InstrumentDetail(InstrumentListItem):
    tick_size: Decimal | None = None
    lot_size: int | None = None
    is_active: bool = True
    quote: QuoteResponse | None = None
    stats: SignalStats = SignalStats()

File: android/domain/src/main/java/com/bofedge/domain/model/Instrument.kt
package com.bofedge.domain.model

/** Instrument reference data (symbol/name/sector are real facts, never prices). */
data class Instrument(
    val id: String,
    val symbol: String,
    val name: String,
    val type: String,        // STOCK | INDEX | COMMODITY | FOREX | CRYPTO
    val exchange: String,
    val currency: String,
    val sectorName: String?,
)

data class SignalStats(
    val totalSignals: Int = 0,
    val bullish: Int = 0,
    val bearish: Int = 0,
    val confirmed: Int = 0,
    val invalidated: Int = 0,
) {
    val hasAny: Boolean get() = totalSignals > 0
}

data class InstrumentQuote(
    val lastPrice: Double?,
    val change: Double?,
    val changePct: Double?,
    val volume: Long?,
    val updatedAt: String?,
)

data class InstrumentDetail(
    val id: String,
    val symbol: String,
    val name: String,
    val type: String,
    val exchange: String,
    val currency: String,
    val sectorName: String?,
    val tickSize: Double?,
    val lotSize: Int?,
    val quote: InstrumentQuote?,
    val stats: SignalStats,
)

data class PageResult<T>(
    val items: List<T>,
    val total: Int,
    val offset: Int,
) {
    val endReached: Boolean get() = offset + items.size >= total
}
